File: reactor/r2dbc-mysql/src/main/java/ua/mai/zyme/r2dbcmysql/exception/FaultDetails.java
package ua.mai.zyme.r2dbcmysql.exception;

import java.text.MessageFormat;
import java.util.Arrays;
import java.util.List;

public class FaultDetails {
    FaultInfo faultInfo;
    List<Object> errorParameters;

    public FaultDetails(FaultInfo faultInfo, Object...errorParameters) {
        this.faultInfo = faultInfo;
        this.errorParameters = (errorParameters != null) ? Arrays.stream(errorParameters).toList() : List.of();
    }

    public String getMessage() {
        return MessageFormat.format(faultInfo.messageTemplate(), errorParameters);
    }

}

File: kotlin/kotlin-r2dbc-mysql/kotlin-r2dbc-mysql-be/src/main/java/ua/mai/zyme/r2dbcmysql/exception/FaultDetails.kt
package ua.mai.zyme.r2dbcmysql.exception

import java.text.MessageFormat
import java.util.*


open class FaultDetails(
    val faultInfo: FaultInfo,
    val errorParameters: List<Any?>
) {

    constructor (faultInfo: FaultInfo, vararg errorParameters: Any?) :
      this (faultInfo, listOf(*errorParameters))

    fun getMessage(): String {
        return MessageFormat.format(faultInfo.messageTemplate(), *errorParameters.toTypedArray())
    }

}